Objects/longobject.c

```diff
@@ -434,8 +446,7 @@ PyLong_FromDouble(double dval)
         dval = -dval;
     }
     frac = frexp(dval, &expo); /* dval = frac*2**expo; 0.0 <= frac < 1.0 */
-    if (expo <= 0)
```

Comment:
This is already on the slow path, it seems safest to keep this check in place even though it should've been handled by the above int range checks.  _smart_ compilers would see that (no idea how many are smart enough to unroll frexp and understand).

Reply:
I do not think it makes sense to keep this code.

Reply:
Either seems fine to me. Personally, I'd probably keep the check out of defensiveness (someone could, for whatever reason, move the fast path out at some point in the future; it's nice if the slow path remains valid in that case), but I'm happy for this to be merged as is. Do we at least have unit tests that cover this case?

Reply:
> smart compilers would see that (no idea how many are smart enough to unroll frexp and understand).

At least gcc is not smart enough.

Reply:
a compromise is to turn it into `assert(expr <= 0);` as protection against future code changes breaking our assumption.  Our buildbots run --with-pydebug builds assertions enabled.

Reply:
It is.